### PR TITLE
ec2_vol: Add check_mode support to ec2_vol

### PR DESCRIPTION
**Pull request by mandar242**
##### SUMMARY

Add check_mode support to ec2_vol.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ec2_vol

### CODE REVIEW

**Comment by goneri** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

File: plugins/modules/ec2_vol.py [hunk 1]
```diff
@@ -809,6 +815,14 @@ def main():
                         changed=False
                     )
 
+        if module.check_mode:
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vol.py:818]: @mandar242  Thank you for working on this. What would be the behaviour if creating or modifying a volume in check mode? At this stage, I don't think it is handled. The check mode should cover all the module's functionality. 

I think this code here could also be removed. It will be better to manage the check mode at function level wether possible.

**Review by alinabuzachis — CHANGES_REQUESTED** [review]

File: changelogs/fragments/509-ec2_vol_add_check_mode_support.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - ec2_vol - add check_mode support
+    (https://github.com/ansible-collections/amazon.aws/pull/509)
```
Comment (alinabuzachis) [inline changelogs/fragments/509-ec2_vol_add_check_mode_support.yml:3]:
```suggestion
    (https://github.com/ansible-collections/amazon.aws/pull/509).
```

File: plugins/modules/ec2_vol.py [hunk 3]
```diff
@@ -696,6 +701,7 @@ def main():
             ['volume_type', 'io1', ['iops']],
             ['volume_type', 'io2', ['iops']],
         ],
+        supports_check_mode=True
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vol.py:704]:
```suggestion
        supports_check_mode=True,
```

**Review by alinabuzachis — CHANGES_REQUESTED** [review]

File: plugins/modules/ec2_vol.py (hunk 1 above)
Comment (mandar242) [reply plugins/modules/ec2_vol.py:818]: Hi @alinabuzachis, Thank you for feedback.
From local testing, I believe this code block handles cases for create/update/detach volume in check_mode. 
Please let me know if you suggest modifying it to more function level, although I think the output would remain same in either case.
[818-824](https://github.com/mandar242/amazon.aws/blob/ec2_vol_check_mode_support/plugins/modules/ec2_vol.py#L818)

Comment (alinabuzachis) [reply plugins/modules/ec2_vol.py:818]: @mandar242 what I was suggesting is instead of stopping here, try to follow the flow and manage the check mode inside create_volume() or update_volume() for example. Similar to what you did in the PR where you implemented the check mode for ec2_snapshot. I also suggest adding some other tests to cover the other functionalities of the module using check mode ( modify a volume, idempotence, etc)

**Comment by alinabuzachis** [comment]
recheck

**Review by jillr — COMMENTED** [review]
Thanks @mandar242. What you have here looks good but could you please add some tests for different update actions? As the code flows now it should be fine, but for example `module_utils.ec2.ensure_ec2_tags` doesn't handle `check_mode` so if the logic ever got changed around we could miss that if there wasn't a test for tag changes under `check_mode`.

**Comment by tremble** [comment]
> `module_utils.ec2.ensure_ec2_tags` doesn't handle `check_mode` so if the logic ever got changed around we could miss that if there wasn't a test for tag changes under `check_mode`.

Not entirely true, it's handled down in remove_ec2_tags and add_ec2_tags which will return "changed" but not make the change in check mode.

**Review by alinabuzachis — COMMENTED** [review]
@mandar242 Thank you very much for working on this. I have only last one suggestion. Could you please use as example the tests run without the check mode and add more tests for the check mode in an analogous way (where it makes sense)? We want to be sure the check mode covers all the functionalities.

**Comment by tremble** [comment]
To follow up on what @alinabuzachis said: where possible, the best way to write the tests is for each 'change' that you make within the tests you run the change 4 times:

1. With check_mode - testing `is changed`
2. Without check_mode - testing `is changed` and the other return values
3. With check_mode again, testing `is not changed`
4. Without check_mode again, testing `is not changed` and the other return values.

This can get rather long and repetitive for the more complex modules, but catches a **lot** of annoying little bugs and when someone's reviewing the code it's much easier to see that things like idempotency and check mode are behaving consistently within the module.  I'll often label each test as 

1. Update _SomeThing_ - check_mode
2. Update _SomeThing_
3. Update _SomeThing_ - idempotency - check_mode
4. Update _SomeThing_ - idempotency

**Comment by tremble** [comment]
recheck

**Comment by jillr** [comment]
recheck

@tremble saw your ping in irc after you'd gone.  That's a really interesting CI failure that I don't know about!  Let's see if it happens again?

**Comment by jillr** [comment]
recheck

**Review by tremble — APPROVED** [review]

**Review by alinabuzachis — APPROVED** [review]
@mandar242 Thank you. LGTM!

File: plugins/modules/ec2_vol.py [hunk 4]
```diff
@@ -457,6 +459,9 @@ def create_volume(module, ec2_conn, zone):
 
     volume = get_volume(module, ec2_conn)
 
+    if module.check_mode:
+        module.exit_json(changed=True, msg='Would have created a volume if not in check mode.')
+
     if volume is None:
```
Comment (tremble) [inline plugins/modules/ec2_vol.py:465]: One minor niggle for the future (not worth blocking this over).  I find it's less likely to lead to weird edge cases if you make your check_mode tests as late as possible.  Here for example, if someone re-wrote the flow of main() and you'd sometimes be passed a volume, performing the check_mode test before you test volume would result in broken idempotency.

**Review by tremble — APPROVED** [review]
